File: code/hashing/custom-hash.java
// A record derives equals/hashCode from its fields; override
// hashCode to combine them. Equal keys must hash equally.
record Point(int x, int y) {
    @Override public int hashCode() {
        return Integer.hashCode(x) * 31 + Integer.hashCode(y);
    }
}
Map<Point, Integer> m = new HashMap<>();

File: code/hashing/custom-hash.py
# Define __hash__ and __eq__ together; hashing a tuple of the
# fields is the idiomatic combine. Equal keys must hash equally.
class Point:
    def __init__(self, x, y):
        self.x, self.y = x, y
    def __hash__(self):
        return hash((self.x, self.y))
    def __eq__(self, other):
        return (self.x, self.y) == (other.x, other.y)

m = {}          # a dict keyed by Point
